### ISSUE TITLE
关于CDDataset的参数为问题请教

### ISSUE THREAD

**Issue by YuQi9797** [PaddlePaddle/PaddleRS#141]
各位大佬您们好，
       我想问下“变化监测任务”是否也有多标签的存在呢？代码参数中`label_list` 应该是用于多标签的任务把？ 请问这一块的数据集是什么样的呢？  又是如何使用训练呢？

**Comment by Bobholamovic** [comment]
你好，PaddleRS对多标签变化检测任务的处理方式近似于语义分割任务，你可以在label list中填写每个类别的名称。
在保证file list满足[PaddleRS 数据集格式要求](https://github.com/PaddlePaddle/PaddleRS/blob/develop/docs/apis/data_cn.md#%E5%8F%98%E5%8C%96%E6%A3%80%E6%B5%8B%E6%95%B0%E6%8D%AE%E9%9B%86cddataset)后，即可像二类变化检测那样执行训练。

**Comment by YuQi9797** [comment]
请问多标签变化检测任务是否是指：
当with_seg_labels为True时，file list中的每一行应该包含5个以空格分隔的项，其中前3项的表示含义与with_seg_labels为False时相同，后2项依次表示第一时相和第二时相影像对应的分割标签相对data_dir的路径。
如果是， 其中后2项&nbsp; 表示第一实相影像分割标签 和 第二实相影像分割标签是什么意思呢？&nbsp; &nbsp;是指有影响变化的地方， 在第一实相影像和第二实相影像都标注出来吗？




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lin ***@***.***&gt;; 
发送时间: 2023年4月21日(星期五) 中午11:06
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [PaddlePaddle/PaddleRS] 关于CDDataset的参数为问题请教 (Issue #141)





 
你好，PaddleRS对多标签变化检测任务的处理方式近似于语义分割任务，你可以在label list中填写每个类别的名称。
 在保证file list满足PaddleRS 数据集格式要求后，即可像二类变化检测那样执行训练。

**Comment by Bobholamovic** [comment]
我想我们可能需要明确一下在你的语境中，『多标签』指的是什么。

机器学习领域的多标签分类（multi-label classification）任务一般指的是一个输入可以同时被分为多类。

> 当with_seg_labels为True时，file list中的每一行应该包含5个以空格分隔的项，其中前3项的表示含义与with_seg_labels为False时相同，后2项依次表示第一时相和第二时相影像对应的分割标签相对data_dir的路径。

这个规则则适用于有『多种类型的标签』的变化检测数据集，比如，数据集中既有变化标签，也有每个时相的建筑物提取标签。从算法的角度上看，这个场景更经常被称为『多任务』。

还有一种对『多标签』的解释是多分类，即有多个类别（二分类只有2个类别），但一个输入只能被分为一类。由于在问题中你提到了label list，当时我倾向于你指的是与label list直接相关的多分类，也有可能是我理解错了，这方面我们可能需要再确认一下~

**Comment by YuQi9797** [comment]
我理解的 在变化检测中的多标签是指：对于两个时相影像中，存在不同地物的变化，对应着不同的标签。&nbsp; 比如一块房子变化了，这块房子为一个标签，&nbsp; 河流变化了， 河流对应一个标签。&nbsp; 即 类似于 实例分割。


那么实际上，咱们这里的多标签在变化检测任务中指的是什么呢？ 如果近似于语义分割任务？&nbsp; 标签的标注是怎么一个标注法呢？ `当with_seg_labels为True时，file list中的每一行应该包含5个以空格分隔的项，其中前3项的表示含义与with_seg_labels为False时相同，后2项依次表示第一时相和第二时相影像对应的分割标签相对data_dir的路径`。&nbsp; 对于后2项，其对应的分割标签是什么意思呢？








------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lin ***@***.***&gt;; 
发送时间: 2023年4月21日(星期五) 晚上7:15
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [PaddlePaddle/PaddleRS] 关于CDDataset的参数为问题请教 (Issue #141)





 
我想我们可能需要明确一下在你的语境中，『多标签』指的是什么。
 
机器学习领域的多标签分类（multi-label classification）任务一般指的是一个输入可以同时被分为多类。
  
当with_seg_labels为True时，file list中的每一行应该包含5个以空格分隔的项，其中前3项的表示含义与with_seg_labels为False时相同，后2项依次表示第一时相和第二时相影像对应的分割标签相对data_dir的路径。
  
这个规则则适用于有『多种类型的标签』的变化检测数据集，比如，数据集中既有变化标签，也有每个时相的建筑物提取标签。从算法的角度上看，这个场景更经常被称为『多任务』。
 
还有一种对『多标签』的解释是多分类，即有多个类别，但一个输入只能被分为一类。由于在问题中你提到了label list，当时我倾向于你指的是与label list直接相关的多分类，也有可能是我理解错了，这方面我们可能需要再确认一下~

**Comment by Bobholamovic** [comment]
这样的话我想你指的应该确实是“多分类”，这种情况file list的每行只需要提供3项内容：两幅输入图像的路径以及标签的路径。标签的标注格式参考语义分割任务，建议使用单通道的png图像，其中存储类别编号（第0类、第1类等等）。

**Comment by YuQi9797** [comment]
那么这里的标签，应该只是一个二值化的mask图吧？ 即记录了哪些地方是有变化的。&nbsp; 但具体是什么地物的变化是不知道的，只知道对应区域是存在变化的？




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lin ***@***.***&gt;; 
发送时间: 2023年4月21日(星期五) 晚上8:01
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [PaddlePaddle/PaddleRS] 关于CDDataset的参数为问题请教 (Issue #141)





 
这样的话我想你指的应该确实是“多分类”，这种情况file list的每行只需要提供3项内容：两幅输入图像的路径以及标签的路径。标签的标注格式参考语义分割任务，建议使用单通道的png图像，其中存储类别编号（第0类、第1类等等）。

**Comment by Bobholamovic** [comment]
如果要做多类变化检测的话，标签就不只是0和1了，比如有0、1、2三类，0表示不变化，1表示建筑物变成了其他地物类型，2表示其他地物类型变成了建筑物。如果需要区分“变化类型”，我理解只靠0、1这两个类别应该是不够的。

**Comment by YuQi9797** [comment]
奥奥奥，就是说多类变化检测的话，标签还是单通道图， 只不过上面的像素值，会因为你的类别数，分别赋值0、1、2等。&nbsp; 对应着三类变化。而如果是单纯看变化与否， 就0，1两类。&nbsp; 像素值对应着0和1 ， 对吧？


然后我还想请教下，就是我看变化检测的代码,用到的数据集是`AirChange`，这里的标签对应代码中的mask，&nbsp; 其值是 0~255.&nbsp; 然后代码中也未对标签进行任何处理。 这里的2分类，就不是指0、1把？&nbsp; &nbsp;而是0 和 255？
https://github.com/PaddlePaddle/PaddleRS/blob/2f6c71ddd2943235100b23eb19dbe053ea5ddfc3/paddlers/transforms/operators.py#L331








------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lin ***@***.***&gt;; 
发送时间: 2023年4月21日(星期五) 晚上8:29
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [PaddlePaddle/PaddleRS] 关于CDDataset的参数为问题请教 (Issue #141)





 
如果要做多类变化检测的话，标签就不只是0和1了，比如有0、1、2三类，0表示不变化，1表示建筑物变成了其他地物类型，2表示其他地物类型变成了建筑物。如果需要区分“变化类型”，我理解只靠0、1这两个类别应该是不够的。

**Comment by Bobholamovic** [comment]
对于变化检测数据集，标签值为0、255的话，可以设置binarize_labels参数为True，这样程序会自动将标签值转换为0、1。

**Comment by YuQi9797** [comment]
奥奥奥，我唐突了。&nbsp; 这里训练代码确实是转化为0、1的了。&nbsp; &nbsp;1.&nbsp; 如果对于多个分类（例如3类）的话，&nbsp; 是不是就自行将其对应区域的像素值转化变为0， 1， 2 呢？


2. https://github.com/PaddlePaddle/PaddleRS/blob/2f6c71ddd2/paddlers/tasks/change_detector.py#L571&nbsp; &nbsp;，对于模型预测输出的结果中，请问下 这里的label_map 是否即为预测的标签mask，其值就是0， 1的呢？&nbsp; 然后score_map表示的是每一像素值位置的一个置信度？





------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lin ***@***.***&gt;; 
发送时间: 2023年4月21日(星期五) 晚上9:20
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [PaddlePaddle/PaddleRS] 关于CDDataset的参数为问题请教 (Issue #141)





 
对于变化检测数据集，标签值为0、255的话，可以设置binarize_labels参数为True，这样程序会自动将标签值转换为0、1。

**Comment by Bobholamovic** [comment]
多类的离散化功能我们暂时还不支持，如果自制数据集的话，建议还是直接标注为0开始的连续数字。

**Comment by YuQi9797** [comment]
奥奥奥，谢谢您的解答！


我还有个疑问，即模型预测后的结果，这里的`label_map`和`score_map`对应的是什么呢？
```
def run(self, net, inputs, mode):
    inputs, batch_restore_list = inputs
    net_out = net(inputs['image'], inputs['image2'])
    logit = net_out[0]
    outputs = OrderedDict()
    if mode == 'test':
        if self.status == 'Infer':
            label_map_list, score_map_list = self.postprocess(
                net_out, batch_restore_list)
        else:
            logit_list = self.postprocess(logit, batch_restore_list)
            label_map_list = []
            score_map_list = []
            for logit in logit_list:
                logit = paddle.transpose(logit, perm=[0, 2, 3, 1])  # NHWC
                label_map_list.append(
                    paddle.argmax(
                        logit, axis=-1, keepdim=False, dtype='int32')
                    .squeeze().numpy())
                score_map_list.append(
                    F.softmax(
                        logit, axis=-1).squeeze().numpy().astype('float32'))
        outputs['label_map'] = label_map_list
        outputs['score_map'] = score_map_list
```
请问这里的 label_map 是否即为模型预测的mask， 只不过这里也是0,1&nbsp; 而不是0,255。
然后score_map的通道数为2，分别表示整HW维度上的像素值，属于0类的置信度， 和属于1的置信度吗？




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lin ***@***.***&gt;; 
发送时间: 2023年4月21日(星期五) 晚上9:34
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [PaddlePaddle/PaddleRS] 关于CDDataset的参数为问题请教 (Issue #141)





 
多类的离散化功能我们暂时还不支持，如果自制数据集的话，建议还是直接标注为0开始的连续数字。

**Comment by Bobholamovic** [comment]
是的，你的理解完全正确。

**Comment by Bobholamovic** [comment]
我们在文档中也做出了解释：https://github.com/PaddlePaddle/PaddleRS/blob/develop/docs/apis/infer_cn.md

**Comment by YuQi9797** [comment]
万分感谢你耐心的解惑！&nbsp; 谢谢谢谢！




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Lin ***@***.***&gt;; 
发送时间: 2023年4月21日(星期五) 晚上10:15
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [PaddlePaddle/PaddleRS] 关于CDDataset的参数为问题请教 (Issue #141)





 
是的，你的理解完全正确。
